### ISSUE TITLE
Follow up on issue 23 fix

### ISSUE THREAD

**Issue by lxylxy123456** [superymk/xmhf64#25]
It looks like https://github.com/superymk/xmhf64/issues/23 is not completely fixed. `MEM_REGION_OVERLAP` and its caller do not check that 64-bit unsigned integer does not overflow.

https://github.com/superymk/xmhf64/commit/40a39f045c1bb5d8b207f8720f4916d448ce0217#diff-1e6192a5c265bb49be8f7c8bb6a74c8816d2738cc9d015b629a89b872ead0413R147-R161

Demo program (compile on 64-bit Linux):
```c
#include <stdio.h>
#include <stdbool.h>

#ifndef __amd64__
#error "Please compile in 64-bit"
#endif

typedef long unsigned ulong_t;
typedef long long unsigned uint64_t;

/// @brief Return if [v1, (v1 + sz1)) overlaps with the region [v2, (v2 + sz2)). MEM_REGION_OVERLAP is architecture 
/// specific because (v1 + sz1) or (v2 + sz2) could equal to 2^32 and hence overflows the uint type of the architecture.
static inline bool MEM_REGION_OVERLAP(ulong_t v1, size_t sz1, ulong_t v2, size_t sz2)
{
    uint64_t base1 = v1;
    uint64_t end1 = base1 + sz1;
    uint64_t base2 = v2;
    uint64_t end2 = base2 + sz2;

    printf("base1 = %016llx\n", base1);
    printf("end1  = %016llx\n", end1);
    printf("base2 = %016llx\n", base2);
    printf("end2  = %016llx\n", end2);

    if((((base1) >= (base2)) && ((base1) < (end2))) || (((base2) >= (base1)) && ((base2) < (end1))))
        return true;

    return false;
}

int main() {
	_Static_assert(sizeof(uint64_t) == 8, "!");
	_Static_assert(sizeof(ulong_t) == 8, "!");
	_Static_assert(sizeof(size_t) == 8, "!");

	printf("%d\n", (int)(MEM_REGION_OVERLAP(
		0xff00000000000003,
		0x00000000dd000001,
		0x8000000000000001,
		0x8000000000000001
	)));

	return 0;
}
```

Output:
```
base1 = ff00000000000003
end1  = ff000000dd000004
base2 = 8000000000000001
end2  = 0000000000000002
0
```

`MEM_REGION_OVERLAP` returns 0, even though the two memory ranges overlap.

**Comment by lxylxy123456** [comment]
Suggested fix:
```c
#include <stdio.h>
#include <stdbool.h>

#ifndef __amd64__
#error "Please compile in 64-bit"
#endif

typedef long unsigned ulong_t;
typedef long long unsigned uint64_t;

#define UINT64_MAX 0xFFFFFFFFULL

// Returns true if add does not overflow.
static inline bool safe_add(uint64_t *out, uint64_t op1, uint64_t op2)
{
	if (op1 <= UINT64_MAX - op2) {
		*out = op1 + op2;
		return true;
	} else {
		return false;
	}
}

// Returns true if [v1, (v1 + sz1)) and [v2, (v2 + sz2)) overlap, or overflow.
static inline bool MEM_REGION_OVERLAP(uint64_t v1, uint64_t sz1, uint64_t v2, uint64_t sz2)
{
	uint64_t base1 = v1;
	uint64_t end1;
	uint64_t base2 = v2;
	uint64_t end2;

	if (sz1 == 0 || sz2 == 0) {
		return false;
	}
	
	if (!safe_add(&end1, base1, sz1 - 1)) {
		return true;
	}

	if (!safe_add(&end2, base2, sz2 - 1)) {
		return true;
	}

	printf("base1 = %016llx\n", base1);
	printf("end1  = %016llx\n", end1);
	printf("base2 = %016llx\n", base2);
	printf("end2  = %016llx\n", end2);

	return base1 <= end2 && base2 <= end1;
}

int main() {
	_Static_assert(sizeof(uint64_t) == 8, "!");
	_Static_assert(sizeof(ulong_t) == 8, "!");
	_Static_assert(sizeof(size_t) == 8, "!");

	printf("%d\n", (int)(MEM_REGION_OVERLAP(
		0xff00000000000003,
		0x00000000dd000001,
		0x8000000000000001,
		0x8000000000000001
	)));

	return 0;
}
```

**Comment by superymk** [comment]
Thanks! Fixed in 592902c8e.
